File: src/main/java/szewek/mcflux/items/ItemMFTool.java
package szewek.mcflux.items;

import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.TextComponentTranslation;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.world.World;
import szewek.mcflux.U;
import szewek.mcflux.api.MCFluxAPI;
import szewek.mcflux.api.ex.Battery;
import szewek.mcflux.api.ex.IEnergy;
import szewek.mcflux.fluxable.WorldChunkEnergy;
import szewek.mcflux.tileentities.TileEntityEnergyMachine;

import javax.annotation.Nonnull;

public class ItemMFTool extends ItemMCFlux {
	private final TextComponentTranslation
			textBlock = new TextComponentTranslation("mcflux.blockcompat.start"),
			textEntity = new TextComponentTranslation("mcflux.entitycompat.start"),
			textIsCompat = new TextComponentTranslation("mcflux.iscompat"),
			textNoCompat = new TextComponentTranslation("mcflux.nocompat"),
			textEnergyUnknown = new TextComponentTranslation("mcflux.energystatunknown"),
			textWorldChunk = new TextComponentTranslation("mcflux.worldchunk");

	public ItemMFTool() {
		setMaxStackSize(1);
		textIsCompat.getStyle().setColor(TextFormatting.GREEN).setBold(true);
		textNoCompat.getStyle().setColor(TextFormatting.RED).setBold(true);
	}

	@Nonnull @Override
	public EnumActionResult onItemUse(ItemStack is, EntityPlayer p, World w, BlockPos pos, EnumHand h, EnumFacing f, float x, float y, float z) {
		if (!w.isRemote) {
			TileEntity te = w.getTileEntity(pos);
			if (te != null) {
				if (te instanceof TileEntityEnergyMachine) {
					TileEntityEnergyMachine teem = (TileEntityEnergyMachine) te;
					if (teem.getModuleId() < 2)
						p.addChatComponentMessage(new TextComponentTranslation("mcflux.transfer", ((TileEntityEnergyMachine) te).getTransferSide(f)));
					return EnumActionResult.SUCCESS;
				}
				IEnergy ie = MCFluxAPI.getEnergySafely(te, f);
				TextComponentTranslation tcb = textBlock.createCopy();
				tcb.appendSibling(ie != null ? textIsCompat : textNoCompat).appendSibling(new TextComponentTranslation("mcflux.blockcompat.end", f));
				p.addChatComponentMessage(tcb);
				if (ie != null)
					p.addChatComponentMessage(new TextComponentTranslation("mcflux.energystat", U.formatMF(ie.getEnergy(), ie.getEnergyCapacity())));
			} else {
				WorldChunkEnergy wce = w.getCapability(WorldChunkEnergy.CAP_WCE, null);
				if (wce != null) {
					Battery bat = wce.getEnergyChunk((int) p.posX, (int) (p.posY + 0.5), (int) p.posZ);
					TextComponentTranslation tcb = textWorldChunk.createCopy();
					tcb.appendSibling(new TextComponentTranslation("mcflux.energystat", U.formatMF(bat.getEnergy(), bat.getEnergyCapacity())));
					p.addChatComponentMessage(tcb);
				} else {
					return EnumActionResult.PASS;
				}
			}
			return EnumActionResult.SUCCESS;
		}
		return EnumActionResult.PASS;
	}

	@Override
	public boolean itemInteractionForEntity(ItemStack is, EntityPlayer p, EntityLivingBase elb, EnumHand h) {
		if (!elb.worldObj.isRemote) {
			IEnergy ie = MCFluxAPI.getEnergySafely(elb, null);
			TextComponentTranslation tcb = textEntity.createCopy();
			tcb.appendSibling(ie != null ? textIsCompat : textNoCompat);
			tcb.appendSibling(new TextComponentTranslation("mcflux.entitycompat.end"));
			p.addChatComponentMessage(tcb);
			if (ie != null) {
				long n = ie.getEnergy(), nc = ie.getEnergyCapacity();
				p.addChatComponentMessage(nc == 1 ? textEnergyUnknown : new TextComponentTranslation("mcflux.energystat", U.formatMF(n, nc)));
			}
			return true;
		}
		return false;
	}
}

File: src/main/java/szewek/mcflux/items/ItemMFTool.kt
package szewek.mcflux.items

import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentTranslation
import net.minecraft.util.text.TextFormatting
import net.minecraft.world.World
import szewek.fl.FLU
import szewek.mcflux.U
import szewek.mcflux.fluxable.FluxableCapabilities
import szewek.mcflux.fluxcompat.FluxCompat
import szewek.mcflux.tileentities.TileEntityEnergyMachine

class ItemMFTool : ItemMCFlux() {
	private val textBlock = TextComponentTranslation("mcflux.blockcompat.start")
	private val textEntity = TextComponentTranslation("mcflux.entitycompat.start")
	private val textIsCompat = TextComponentTranslation("mcflux.iscompat")
	private val textNoCompat = TextComponentTranslation("mcflux.nocompat")
	private val textEnergyUnknown = TextComponentTranslation("mcflux.energystatunknown")
	private val textWorldChunk = TextComponentTranslation("mcflux.worldchunk")

	init {
		setMaxStackSize(1)
		textIsCompat.style.setColor(TextFormatting.GREEN).bold = true
		textNoCompat.style.setColor(TextFormatting.RED).bold = true
	}

	override fun onItemUse(p: EntityPlayer, w: World, pos: BlockPos, h: EnumHand, f: EnumFacing, x: Float, y: Float, z: Float): EnumActionResult {
		if (!w.isRemote) {
			val te = w.getTileEntity(pos)
			if (te != null) {
				if (te is TileEntityEnergyMachine) {
					val teem = te as TileEntityEnergyMachine?
					if (teem!!.moduleId < 2)
						p.sendMessage(TextComponentTranslation("mcflux.transfer", te.getTransferSide(f)))
					return EnumActionResult.SUCCESS
				}
				val ie = FLU.getEnergySafely(te, f)
				val tcb = textBlock.createCopy()
				tcb.appendSibling(if (ie != null) textIsCompat else textNoCompat).appendSibling(TextComponentTranslation("mcflux.blockcompat.end", f))
				p.sendMessage(tcb)
				if (ie != null)
					p.sendMessage(TextComponentTranslation("mcflux.energystat", U.formatMF(ie)))
				if (ie is FluxCompat.Convert)
					p.sendMessage(TextComponentTranslation("mcflux.fluxcompat.type", (ie as FluxCompat.Convert).energyType.name))

			} else {
				val wce = w.getCapability(FluxableCapabilities.CAP_WCE, null)
				if (wce != null) {
					val bat = wce.getEnergyChunk(p.posX.toInt(), (p.posY + 0.5).toInt(), p.posZ.toInt())
					val tcb = textWorldChunk.createCopy()
					tcb.appendSibling(TextComponentTranslation("mcflux.energystat", U.formatMF(bat)))
					p.sendMessage(tcb)
				} else {
					return EnumActionResult.PASS
				}
			}
			return EnumActionResult.SUCCESS
		} else if (p.isSneaking) {
			return EnumActionResult.SUCCESS
		}
		return EnumActionResult.PASS
	}

	override fun itemInteractionForEntity(stk: ItemStack, p: EntityPlayer, elb: EntityLivingBase, h: EnumHand): Boolean {
		if (!elb.world.isRemote) {
			val ie = FLU.getEnergySafely(elb, EnumFacing.UP) // FIXME: SET TO null
			val tcb = textEntity.createCopy()
			tcb.appendSibling(if (ie != null) textIsCompat else textNoCompat)
			tcb.appendSibling(TextComponentTranslation("mcflux.entitycompat.end"))
			p.sendMessage(tcb)
			if (ie != null) {
				val nc = ie.energyCapacity
				p.sendMessage(if (nc == 1L) textEnergyUnknown else TextComponentTranslation("mcflux.energystat", U.formatMF(ie)))
			}
			if (ie is FluxCompat.Convert)
				p.sendMessage(TextComponentTranslation("mcflux.fluxcompat.type", (ie as FluxCompat.Convert).energyType.name))
			return true
		}
		return false
	}
}
